### PR TITLE
elmerfem: make elmerfem-9.0 run properly

### PR DESCRIPTION
**Pull request by grindhold**
###### Description of changes

This is the result of my efforts to clear out the problems mentioned in #167599 . The import-problem was not the only one to fix in order to receive a working build of ElmerFEM:

  0) ElmerGUI displays things in the Model tree on the left screen side now [Before it was blank space which is not as intended by the developers]
  1) Enable the import of STEP- and other format by building against opencascade [This is used in many tutorials and will annoy users even if the software is theoretically usable without it]
  2) Patch the source to enable builds with gcc-11 and gfortran-11
  3) Build against Qt5  instead of Qt4 and supply relevant patches

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 1]
```diff
@@ -13,19 +13,42 @@ stdenv.mkDerivation rec {
...
-  buildInputs = [ mpi blas liblapack qt4 qwt6_qt4 ];
+  nativeBuildInputs = [
+    cmake
+    gfortran
+    pkg-config
+    git
+    qt5.wrapQtAppsHook
+  ];
+  buildInputs = [
+    mpi blas liblapack
```
Comment (Wulfsta) [inline pkgs/applications/science/physics/elmerfem/default.nix:24]: Can you separate these onto their own lines?

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 2]
```diff
@@ -13,19 +13,42 @@ stdenv.mkDerivation rec {
...
+    libsForQt5.qt5.qtbase
+    libsForQt5.qt5.qtscript
+    libsForQt5.qwt
+    libGL libGLU
```
Comment (Wulfsta) [inline pkgs/applications/science/physics/elmerfem/default.nix:28]: See above.

**Review by Wulfsta — CHANGES_REQUESTED** [review]
Looks good other than formatting!

**Comment by grindhold** [comment]
i applied your one-dependency-per-line suggestion but i think i messed up the rebase. will fix tomorrowb

**Comment by grindhold** [comment]
fixed the rebase

**Review by Wulfsta — APPROVED** [review]

**Comment by grindhold** [comment]
Is there anything left to do in order to get this merged?

**Comment by Wulfsta** [comment]
Hmm, there’s a check that’s still running? Maybe try rerunning those?

**Comment by grindhold** [comment]
i was told by a community member that the aarch64-darwin builldboxes were broken lately and have a huge backlog now.

**Comment by Wulfsta** [comment]
It’s probably just going to take a little while to get picked up by someone with merge permissions then.

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 3]
```diff
@@ -13,19 +13,45 @@ stdenv.mkDerivation rec {
 
   hardeningDisable = [ "format" ];
 
-  nativeBuildInputs = [ cmake gfortran pkg-config git ];
-  buildInputs = [ mpi blas liblapack qt4 qwt6_qt4 ];
+  nativeBuildInputs = [
+    cmake
+    gfortran
+    pkg-config
+    git
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:20]: Why do we need git?

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 4]
```diff
@@ -13,19 +13,45 @@ stdenv.mkDerivation rec {
...
+    libsForQt5.qt5.qtscript
+    libsForQt5.qwt
+    libGL
+    libGLU
+    opencascade
+    vtkWithQt5
+  ];
 
   preConfigure = ''
     patchShebangs ./
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:37]: Can we be more precise?

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 5]
```diff
@@ -13,19 +13,45 @@ stdenv.mkDerivation rec {
...
+    pkg-config
+    git
+    qt5.wrapQtAppsHook
+  ];
+  buildInputs = [
+    mpi
+    blas
+    liblapack
+    libsForQt5.qt5.qtbase
+    libsForQt5.qt5.qtscript
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:28]:
```suggestion
    libsForQt5.qtbase
    libsForQt5.qtscript
```

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, cmake, git, gfortran, mpi, blas, liblapack, qt4, qwt6_qt4, pkg-config }:
+{ lib, stdenv, fetchFromGitHub, cmake, git, gfortran, mpi, blas, liblapack, qt5, pkg-config, libGL, libGLU, opencascade, libsForQt5, vtkWithQt5}:
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:1]: Either use qt5 or libForQt5.

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 3 above)
Comment (Wulfsta) [reply pkgs/applications/science/physics/elmerfem/default.nix:20]: We probably don’t, looks like an oversight from when I originally wrote this.

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 4 above)
Comment (Wulfsta) [reply pkgs/applications/science/physics/elmerfem/default.nix:37]: As I recall there are many files to patch, I don’t think this is an issue.

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 6 above)
Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix]: done

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 5 above)
Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix]: done

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 3 above)
Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix]: done

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/applications/science/physics/elmerfem/default.nix:36]: > I don’t think this is an issue.

Well, it takes longer if the script looks at every file.

**Pull request merged**